docs/CONFIG_REFERENCE.md
# Config Reference

[Back](../README.md)

This document contains the explaination for all the different values in the config.json file

## External address (externalAddress)

This address should be an ip or hostname which points to the server that is hosting KME
this address must be reachable by the clients connecting, so it must be either a valid IP
address or a hostname that points to it.

The default for this is 383933-gosprapp396.ea.com when the default value is used its required
that this value be redirected in the users hosts file

## Ports  (ports)

The values in this object represents the differents ports that each of the child
servers should be bound on. These port values must be integer values

| Key        | What                                                                                                                                                                      |
|------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| redirector | This is the port which the redirector server is bound to                                                                                                                  |
| main       | This is the port which the main game server is bound to                                                                                                                   |
| discard    | This is the port that the discard server is bound to this port is not very important, so it can be given a low priority number as anything sent to it is simply discarded |
| http       | This is the port of the http server which both the panel and the game images are hosted on                                                                                |

> NOTE: You should **NOT** change the port of the redirector server to anything
> other than the default 42127 unless you are behind a proxy or inside a container
> that is forwarding from 42127 beacuse this is the port that the mass effect client
> trys to make a connection to

## Logging (logging)

This is the configuration for the application logging this defines whichs levels of logging
should be used as well as other settings for logging

| Key     | What                                                                                                                |
|---------|---------------------------------------------------------------------------------------------------------------------|
| level   | This is the logging level which should be used (INFO, WARN, ERROR, FATAL, DEBUG)                                    |
| save    | This option determines whether log files will be saved in log files                                                 |
| packets | This options determines whether the information of incomming and outgoing packets should be printed to the console. |

> NOTE: I recommend disabling packets logging when using MITM mode because MITM logs packets
> as well which will **GREATLY** clog up your console with confusing messages

## Menu Message (menuMessage)

This is the message displayed on the main menu you can use special formatting codes
here in order to include dynamic information which is replaced by the server

| Code | What                                         |
|------|----------------------------------------------|
| {v}  | Replaced with the version of the KME server  |
| {n}  | Replaced with the authenticated players name |
| {ip} | Replaced with the connections IP address     |

## Database (database)

This configuration section contains the database information that determines which database to use
as well as the credentials

| Key      | What                                                                                        |
|----------|---------------------------------------------------------------------------------------------|
| type     | This is the type of database that should be used (MySQL, or SQLite) name is case-insensitve |
| host     | The host of the database server *(MySQL Only)*                                              |
| port     | The port of the database server  *(MySQL Only)*                                             |
| user     | The username for the database server *(MySQL Only)*                                         |
| password | The password for the database server *(MySQL Only)*                                         |
| database | The name of the database *(MySQL Only)*                                                     |
| file     | The database file path *(SQLite Only)*  This path is from the current working directory     |

## Panel (panel)

This configuration section contains the settings for the HTTP web panel

| Key      | What                                 |
|----------|--------------------------------------|
| enabled  | Whether the panel is enabled         |
| username | The username for accessing the panel |
| password | The password for accessing the panel |

## Galaxy at war (gaw)

This configuration contains the galaxy at war settings

| Key                 | What                                                                                                                  |
|---------------------|-----------------------------------------------------------------------------------------------------------------------|
| readinessDailyDecay | The amount of readiness level to decay each day from last update 0.5 = -1%. Set this value defaults to 0 for no decay |
| enablePromotions    | Whether to count promotions on the galaxy at war                                                                      |

## Man In The Middle (mitm)

This configuration section contains the settings for the Man-in-the-middle mode which creates a connection
to the official server and tunnels the connections throught itself decoding and logging all the traffic

| Key     | What                                                                                                                                                |
|---------|-----------------------------------------------------------------------------------------------------------------------------------------------------|
| enabled | Whether the Man-in-the-middle mode should be used                                                                                                   |
| address | The ip address that gosredirector.ea.com points to. If the server doesn't have a hosts file redirect for gosredirector.ea.com then you can use that |


app/src/main/kotlin/com/jacobtread/kme/Config.kt
package com.jacobtread.kme

import kotlinx.serialization.Serializable

/**
 * Config the configuration for this application
 *
 * @property externalAddress The address that clients will use to connect to this server. This address should point to
 * the machine that this server is hosted on and must be accessible to the clients. If you aren't using a custom
 * domain then the default domain should be added to the system hosts file
 * @property ports The different ports that all the different servers should use
 * @property logging The configuration for the logger
 * @property menuMessage The message that should be displayed on the main menu
 * @property database The database configuration
 * @property panel The configuration for the web panel authentication
 * @property gaw The Galaxy At War configuration
 * @constructor Create empty Config
 */
@Serializable
data class Config(
    val externalAddress: String = "383933-gosprapp396.ea.com",
    val ports: Ports = Ports(),
    val logging: LoggingConfig = LoggingConfig(),
    val menuMessage: String = "<font color='#B2B2B2'>KME3</font> - <font color='#FFFF66'>Logged as: {n}</font>",
    val database: DatabaseConfig = DatabaseConfig(),
    val panel: PanelConfig = PanelConfig(),
    val gaw: GalaxyAtWarConfig = GalaxyAtWarConfig(),
    val mitm: Boolean = false,
) {


    /**
     * DatabaseConfig Stores configuration information about the database
     *
     * @property type Defines which database connection type to use
     * @property mysql The config for a MySQL database connection
     * @property sqlite The config for a SQLite database connection
     * @constructor Create empty DatabaseConfig
     */
    @Serializable
    data class DatabaseConfig(
        val type: String = "sqlite",
        val host: String = "127.0.0.1",
        val port: Int = 3306,
        val user: String = "root",
        val password: String = "password",
        val database: String = "kme",
        val file: String = "data/app.db",
    )

    @Serializable
    data class LoggingConfig(
        val level: String = "info",
        val save: Boolean = true,
        val packets: Boolean = false,
    )


    /**
     * Ports Configuration for the different ports that the servers will use
     *
     * NOTE: Telemetry and ticker servers have been merged into one single
     * discard server because they are unused. Original ports were
     * ticker = 8999, telemetry = 9988
     *
     * @property redirector The port for the redirector server
     * @property main The port for the main server
     * @property discard The port for the ticker & telemetry server
     * @property http The port for the http server
     * @constructor Create empty Ports
     */
    @Serializable
    data class Ports(
        val redirector: Int = 42127,
        val main: Int = 14219,
        val discard: Int = 9988,
        val http: Int = 80,
    )


    /**
     * GalaxyAtWarConfig The configuration of the Galaxy at War
     *
     * @property readinessDailyDecay The amount of readiness level to decay each day from last update 0.5 = -1%
     * Set this value defaults to 0 for no decay
     * @property enablePromotions Whether to include promotions in the galaxy at war level
     * @constructor Create empty GalaxyAtWarConfig
     */
    @Serializable
    data class GalaxyAtWarConfig(
        val readinessDailyDecay: Float = 0f,
        val enablePromotions: Boolean = true,
    )

    /**
     * PanelConfig The configuration for the web panel
     *
     * @property enabled Whether the panel is enabled or not
     * @property username The panel username
     * @property password The panel password
     * @constructor Create empty PanelConfig
     */
    @Serializable
    data class PanelConfig(
        val enabled: Boolean = false,
        val username: String = "admin",
        val password: String = "admin",
    )
}